lib/iris/tests/unit/util/test__slice_data_with_keys.py

```diff
@@ -0,0 +1,257 @@
+# (C) British Crown Copyright 2017, Met Office
+#
+# This file is part of Iris.
+#
+# Iris is free software: you can redistribute it and/or modify it under
+# the terms of the GNU Lesser General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Iris is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Iris.  If not, see <http://www.gnu.org/licenses/>.
+"""
+Test function :func:`iris.util._slice_data_with_keys`.
+
+Note: much of the functionality really belongs to the other routines,
+:func:`iris.util._build_full_slice_given_keys`, and
+:func:`column_slices_generator`.
+However, it is relatively simple to test multiple aspects of all three here
+in combination.
+
+"""
+from __future__ import (absolute_import, division, print_function)
+from six.moves import (filter, input, map, range, zip)  # noqa
+
+# import iris tests first so that some things can be initialised before
+# importing anything else
+import iris.tests as tests
+
+import numpy as np
+
+from iris.util import _slice_data_with_keys
+
+
+class DummyArray(object):
+    # A dummy array-like that records the keys of indexing calls.
+    def __init__(self, shape, _indexing_record_list=None):
+        self.shape = shape
+        self.ndim = len(shape)
+        if _indexing_record_list is None:
+            _indexing_record_list = []
+        self._getitem_call_keys = _indexing_record_list
+
+    def __getitem__(self, keys):
+        # Add the indexing keys to the call list.
+        self._getitem_call_keys.append(keys)
+        # Return a new object with the correct derived shape, and record its
+        # indexing operations in the same key list as this.
+        shape_array = np.zeros(self.shape)
+        shape_array = shape_array.__getitem__(keys)
+        new_shape = shape_array.shape
+        return DummyArray(new_shape,
+                          _indexing_record_list=self._getitem_call_keys)
+
+
+class Indexer(object):
+    # An object to make __getitem__ arglists from indexing operations.
+    def __getitem__(self, keys):
+        return keys
+
+
+# An Indexer object for generating indexing keys in a nice visible way.
+Index = Indexer()
+
+
+class MixinIndexingTest(object):
+    def check(self, shape, keys, expect_call_keys=None, expect_map=None):
```

Comment:
> what is really being tested is the behaviour of `column_slices_generator`

Or the combination with `_build_full_slice_given_keys`. See also the [note in the docstring](https://github.com/pp-mo/iris/blob/daf80ac690daabb4cc3fe9fb291a5a5fd46f49e1/lib/iris/tests/unit/util/test__slice_data_with_keys.py#L20-L24) for this test module.

> these tests shouldn't really be necessary

We've introduced a new, untested function. Of course there should be tests for it! Especially when the real driver for these tests was the terrible lack of any other testing for anything within this functionality space. Thus adding these tests is a positive step as it gives us some assurance for the first time that this functionality is doing what we thought it was...

You could perhaps argue that these should be integration tests, or alternatively you could say we're testing the unit that is `iris.util._slice_data_with_keys`.